staging/src/k8s.io/apimachinery/pkg/runtime/extension.go

```diff
@@ -18,10 +18,52 @@ package runtime
 
 import (
 	"bytes"
-	"encoding/json"
 	"errors"
+	"fmt"
+
+	cbor "k8s.io/apimachinery/pkg/runtime/serializer/cbor/direct"
+	"k8s.io/apimachinery/pkg/util/json"
 )
 
+// RawExtension intentionally avoids implementing value.UnstructuredConverter for now because the
+// signature of ToUnstructured does not allow returning an error value in cases where the conversion
+// is not possible (content type is unrecognized or bytes don't match content type).
+func (re RawExtension) toUnstructured(rawContentType string) (interface{}, error) {
+	if re.Raw != nil {
+		switch rawContentType {
+		case ContentTypeJSON:
+			var u interface{}
+			if err := json.Unmarshal(re.Raw, &u); err != nil {
+				return nil, fmt.Errorf("failed to parse RawExtension bytes as JSON: %w", err)
+			}
+			return u, nil
+		case ContentTypeCBOR:
+			var u interface{}
+			if err := cbor.Unmarshal(re.Raw, &u); err != nil {
+				return nil, fmt.Errorf("failed to parse RawExtension bytes as CBOR: %w", err)
+			}
+			return u, nil
+		default:
+			return nil, fmt.Errorf("cannot convert RawExtension with unrecognized content type to unstructured")
+		}
+	}
+	return nil, nil
```

Comment:
Why can't (or shouldn't) we use `DefaultUnstructuredConverter.ToUnstructured` with the `re.Object`?

Reply:
Initially, this implemented https://pkg.go.dev/sigs.k8s.io/structured-merge-diff/v5/value#UnstructuredConverter and it did fall back to Object when Raw was nil (as MarshalJSON does). The signature doesn't allow returning errors, and not all RawExtensions can be automatically converted to Unstructured, so it was changed. Now this function is only used for transcoding in `MarshalJSON` and `MarshalCBOR` when Raw is already known to be non-nil.

I removed the remaining nil check, removed the receiver, and changed the name and signature to `func rawToUnstructured(raw []byte, contentType string) (interface{}, error)` to reflect the current purpose of this function.